### PR TITLE
Ctpdev0

### PR DESCRIPTION
**Pull request by mbombara**
Very first version of the QC CTP - the code can read raw data file and publish BC, trigger Inputs and trigger classes distributions.

### CODE REVIEW

File: Modules/CTP/basic-ctp.json [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        "moduleName": "QcCTP",
+        "detectorName": "CTP",
+        "cycleDurationSeconds": "10",
+        "maxNumberCycles": "-1",
+        "": "The other type of dataSource is \"direct\", see basic-no-sampling.json.",
+        "dataSource": {
+          "type": "dataSamplingPolicy",
+          "name": "tst-raw"
+        },
+        "taskParameters": {
```
Comment (Barthelemy) [inline Modules/CTP/basic-ctp.json:38]: you can probably remove this example parameter

File: Modules/CTP/basic-ctp.json [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+          "myOwnKey": "myOwnValue"
+        },
+        "location": "remote",
+        "saveObjectsToFile": "test.root",      "": "~/alice/QualityControl/Modules/CTP/"
```
Comment (Barthelemy) [inline Modules/CTP/basic-ctp.json:42]: please remove the filename and just add it to your local clone

File: Modules/CTP/basic-ctp.json [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        }]
+      }
+    }
+  },
+  "dataSamplingPolicies": [
+    {
+      "id": "tst-raw",
+      "active": "true",
+      "machines": [],
+      "query": "random:TST/RAWDATA/0",
```
Comment (Barthelemy) [inline Modules/CTP/basic-ctp.json:65]: Are you sending your data under `TST/RAWDATA` ? shouldn't it be under something like `CTP/RAWDATA` ?

File: Modules/CTP/src/RawDataQcCheck.cxx [hunk 4]
```diff
@@ -0,0 +1,85 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
```
Comment (Barthelemy) [inline Modules/CTP/src/RawDataQcCheck.cxx:1]: It seems that RawDataQcCheck is the default skeleton class. Please remove it.

File: Modules/CTP/include/CTP/RawDataQcTask.h [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+class TH1F;
+
+using namespace o2::quality_control::core;
+
+namespace o2::quality_control_modules::ctp
+{
+
+/// \brief Example Quality Control DPL Task
+/// \author My Name
+class RawDataQcTask final : public TaskInterface
```
Comment (Barthelemy) [inline Modules/CTP/include/CTP/RawDataQcTask.h:31]: Please name the class in a more meaningful way.

File: Modules/CTP/src/RawDataQcTask.cxx [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+  delete mHistoInputs;
+  delete mHistoClasses;
+}
+
+void RawDataQcTask::initialize(o2::framework::InitContext& /*ctx*/)
+{
+  ILOG(Info, Support) << "initialize RawDataQcTask" << ENDM; // QcInfoLogger is used. FairMQ logs will go to there as well.
+
+  // this is how to get access to custom parameters defined in the config file at qc.tasks.<task_name>.taskParameters
+  if (auto param = mCustomParameters.find("myOwnKey"); param != mCustomParameters.end()) {
```
Comment (Barthelemy) [inline Modules/CTP/src/RawDataQcTask.cxx:44]: please remove example code

File: Modules/CTP/src/RawDataQcTask.cxx [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+  }
+
+  mHistoBC = new TH1F("histobc", "BC distribution", 3564, 0, 3564);
+  mHistoInputs = new TH1F("inputs", "Inputs distribution", 48, 0, 48);
+  mHistoClasses = new TH1F("classes", "Classes distribution", 64, 0, 64);
+  getObjectsManager()->startPublishing(mHistoBC);
+  getObjectsManager()->startPublishing(mHistoInputs);
+  getObjectsManager()->startPublishing(mHistoClasses);
+  try {
+    getObjectsManager()->addMetadata(mHistoBC->GetName(), "custom", "34");
```
Comment (Barthelemy) [inline Modules/CTP/src/RawDataQcTask.cxx:55]: this is not necessary

File: dpl-config.json [hunk 8]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (Barthelemy) [inline dpl-config.json:1]: is this file needed ?

**Review by Barthelemy — CHANGES_REQUESTED** [review]
Thank you for the new ctp module. Could you review my comments and make the changes ? Then I will re-review and approve.

File: Modules/CTP/basic-ctp.json (hunk 3 above)
Comment (mbombara) [reply Modules/CTP/basic-ctp.json:65]: I don't sample data (yet) so I left it as it was from example. I am changing tst to ctp in both cases.

File: Modules/CTP/basic-ctp.json (hunk 2 above)
Comment (mbombara) [reply Modules/CTP/basic-ctp.json:42]: I have deleted whole row.

File: Modules/CTP/src/RawDataQcCheck.cxx (hunk 4 above)
Comment (mbombara) [reply Modules/CTP/src/RawDataQcCheck.cxx:1]: Yes it is. Do you mean removing RawDataQcCheck.cxx and RawDataQcCheck.h ?

File: Modules/CTP/include/CTP/RawDataQcTask.h (hunk 5 above)
Comment (mbombara) [reply Modules/CTP/include/CTP/RawDataQcTask.h:31]: Would CTPRawDataReaderTask be better?

File: Modules/CTP/src/RawDataQcTask.cxx (hunk 6 above)
Comment (mbombara) [reply Modules/CTP/src/RawDataQcTask.cxx:44]: The problem is I do not know what is artefact and what is really needed. :-) I am removing lines 43 and 44.

File: Modules/CTP/src/RawDataQcTask.cxx (hunk 7 above)
Comment (mbombara) [reply Modules/CTP/src/RawDataQcTask.cxx:55]: I do not know what those lines do, I thought they might be important. Is it ok to delete whole try { } catch { } section (lines 54-65)?

File: dpl-config.json (hunk 8 above)
Comment (mbombara) [reply dpl-config.json:1]: I have no idea, I do not even know how it got there. I have removed it.

File: Modules/CTP/src/RawDataQcCheck.cxx (hunk 4 above)
Comment (Barthelemy) [reply Modules/CTP/src/RawDataQcCheck.cxx:1]: yes, removing the whole class if you don't use it .

File: Modules/CTP/include/CTP/RawDataQcTask.h (hunk 5 above)
Comment (Barthelemy) [reply Modules/CTP/include/CTP/RawDataQcTask.h:31]: yes

File: Modules/CTP/src/RawDataQcTask.cxx (hunk 7 above)
Comment (Barthelemy) [reply Modules/CTP/src/RawDataQcTask.cxx:55]: yes

**Comment by Barthelemy** [comment]
do not forget to push your changes :)

File: Modules/CTP/src/RawDataQcCheck.cxx (hunk 4 above)
Comment (mbombara) [reply Modules/CTP/src/RawDataQcCheck.cxx:1]: also to remove the check in .json file?

Comment (mbombara) [reply Modules/CTP/src/RawDataQcCheck.cxx:1]: and line in LinkDef.h?

Comment (Barthelemy) [reply Modules/CTP/src/RawDataQcCheck.cxx:1]: yes to all

File: Modules/CTP/basic-ctp.json (hunk 3 above)
Comment (Barthelemy) [reply Modules/CTP/basic-ctp.json]: but if you are not sampling, what data are you monitoring ?

**Review by Barthelemy — APPROVED** [review]

File: Modules/CTP/basic-ctp.json (hunk 3 above)
Comment (mbombara) [reply Modules/CTP/basic-ctp.json]: This version is reading raw files locally - i.e. raw file is created from simulation (then digits, then raw file). I wanted to have it while waiting for interface by Sylvain. The payload structure will be the same in the real data so it is not waste of time. If I correctly understand, the data sampling is needed when you want monitor only fraction of data? here I monitor all (they are from simulation).
